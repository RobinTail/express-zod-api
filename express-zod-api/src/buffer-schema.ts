import { z } from "zod/v4";

export const bufferSchema = z.custom<Buffer>(
  (subject) => Buffer.isBuffer(subject),
  { error: "Expected Buffer" },
);
