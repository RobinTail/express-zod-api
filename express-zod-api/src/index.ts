import "@express-zod-api/zod-plugin"; // side effects here
export { createConfig } from "./config-type.ts";
export {
  EndpointsFactory,
  defaultEndpointsFactory,
  arrayEndpointsFactory,
} from "./endpoints-factory.ts";
export { getMessageFromError } from "./common-helpers.ts";
export { ensureHttpError } from "./result-helpers.ts";
export { BuiltinLogger } from "./builtin-logger.ts";
export { Middleware } from "./middleware.ts";
export {
  ResultHandler,
  defaultResultHandler,
  arrayResultHandler,
} from "./result-handler.ts";
export { DependsOnMethod } from "./depends-on-method.ts";
export { ServeStatic } from "./serve-static.ts";
export { createServer, attachRouting } from "./server.ts";
export { Documentation } from "./documentation.ts";
export {
  DocumentationError,
  RoutingError,
  OutputValidationError,
  InputValidationError,
  MissingPeerError,
} from "./errors.ts";
export { testEndpoint, testMiddleware } from "./testing.ts";
export { Integration } from "./integration.ts";
export { EventStreamFactory } from "./sse.ts";

export { ez } from "./proprietary-schemas.ts";

// Convenience types
export type { Depicter } from "./documentation-helpers.ts";
export type { Producer } from "./zts-helpers.ts";

// Interfaces exposed for augmentation
export type { LoggerOverrides } from "./logger-helpers.ts";
export type { TagOverrides } from "./common-helpers.ts";

// Issues 952, 1182, 1269: Insufficient exports for consumer's declaration
import type {} from "qs"; // fixes TS2742 for attachRouting
export type { Routing } from "./routing.ts";
export type { FlatObject } from "./common-helpers.ts";
export type { Method } from "./method.ts";
export type { IOSchema } from "./io-schema.ts";
export type { CommonConfig, AppConfig, ServerConfig } from "./config-type.ts";
export type { ApiResponse } from "./api-response.ts";
export type {
  BasicSecurity,
  BearerSecurity,
  CookieSecurity,
  HeaderSecurity,
  InputSecurity,
  OAuth2Security,
  OpenIdSecurity,
} from "./security.ts";
