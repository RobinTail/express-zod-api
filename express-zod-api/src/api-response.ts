import { z } from "zod";
import { isSchema } from "./common-helpers";
import { contentTypes } from "./content-type";

export const defaultStatusCodes = {
  positive: 200,
  negative: 400,
} satisfies Record<string, number>;

export type ResponseVariant = keyof typeof defaultStatusCodes;
export const responseVariants = Object.keys(
  defaultStatusCodes,
) as ResponseVariant[];

/** @public this is the user facing configuration */
export class ApiResponse<S extends z.ZodType> {
  readonly #schema: S;
  readonly #statusCodes?: [number, ...number[]];
  readonly #mimeTypes?: [string, ...string[]] | null;

  public constructor(
    subject:
      | S // shorthand for { schema: S }, default status codes and MIME types applied
      | {
          schema: S;
          /** @default 200 for a positive and 400 for a negative response */
          statusCode?: number | [number, ...number[]];
          /**
           * @example null is for no content, such as 204 and 302
           * @default "application/json"
           * */
          mimeType?: string | [string, ...string[]] | null;
        },
  ) {
    const { schema, statusCode, mimeType } = isSchema(subject)
      ? { schema: subject }
      : subject;
    this.#schema = schema;
    this.#statusCodes =
      typeof statusCode === "number" ? [statusCode] : statusCode;
    this.#mimeTypes = typeof mimeType === "string" ? [mimeType] : mimeType;
  }

  public normalize(fbStatus: number) {
    return {
      schema: this.#schema,
      statusCodes: this.#statusCodes ?? [fbStatus],
      mimeTypes:
        this.#mimeTypes === undefined ? [contentTypes.json] : this.#mimeTypes,
    };
  }
}

export type NormalizedResponse = ReturnType<
  ApiResponse<z.ZodType>["normalize"]
>;
