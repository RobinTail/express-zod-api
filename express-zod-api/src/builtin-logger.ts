import ansis from "ansis";
import { inspect } from "node:util";
import { performance } from "node:perf_hooks";
import { FlatObject, isProduction } from "./common-helpers";
import {
  AbstractLogger,
  formatDuration,
  isHidden,
  Severity,
  styles,
} from "./logger-helpers";

interface Context extends FlatObject {
  requestId?: string;
}

export interface BuiltinLoggerConfig {
  /**
   * @desc The minimal severity to log or "silent" to disable logging
   * @example "debug" also enables pretty output for inspected entities
   * */
  level: "silent" | "warn" | "info" | "debug";
  /** @desc Enables colors on printed severity and inspected entities */
  color: boolean;
  /**
   * @desc Printing logs does not delay execution
   * @see https://expressjs.com/en/advanced/best-practice-performance.html#do-logging-correctly
   * */
  async: boolean;
  /**
   * @desc Control how deeply entities should be inspected
   * @example null
   * @example Infinity
   * */
  depth: number | null;
  /**
   * @desc Context: the metadata applicable for each logged entry, used by .child() method
   * @see childLoggerProvider
   * */
  ctx: Context;
}

interface ProfilerOptions {
  message: string;
  /** @default "debug" */
  severity?: Severity | ((ms: number) => Severity);
  /** @default formatDuration - adaptive units and limited fraction */
  formatter?: (ms: number) => string | number;
}

/** @desc Built-in console logger with optional colorful inspections */
export class BuiltinLogger implements AbstractLogger {
  protected readonly config: BuiltinLoggerConfig;
  protected readonly stack: string[] = [];
  protected postponed?: NodeJS.Immediate;

  /** @example new BuiltinLogger({ level: "debug", color: true, depth: 4 }) */
  public constructor({
    color = ansis.isSupported(),
    level = isProduction() ? "warn" : "debug",
    async = isProduction(),
    depth = 2,
    ctx = {},
  }: Partial<BuiltinLoggerConfig> = {}) {
    this.config = { color, level, async, depth, ctx };
  }

  protected format(subject: unknown) {
    const { depth, color: colors, level } = this.config;
    return inspect(subject, {
      depth,
      colors,
      breakLength: level === "debug" ? 80 : Infinity,
      compact: level === "debug" ? 3 : true,
    });
  }

  protected print(method: Severity, message: string, meta?: unknown) {
    const {
      level,
      ctx: { requestId, ...ctx },
      color: hasColor,
      async: isAsync,
    } = this.config;
    if (level === "silent" || isHidden(method, level)) return;
    const output: string[] = [new Date().toISOString()];
    if (requestId) output.push(hasColor ? styles.ctx(requestId) : requestId);
    output.push(
      hasColor ? `${styles[method](method)}:` : `${method}:`,
      message,
    );
    if (meta !== undefined) output.push(this.format(meta));
    if (Object.keys(ctx).length > 0) output.push(this.format(ctx));
    (isAsync ? this.postpone.bind(this) : console.log)(output.join(" "));
  }

  protected postpone(line: string) {
    this.stack.push(line);
    clearImmediate(this.postponed);
    this.postponed = setImmediate(this.purge.bind(this)).unref(); // do not block process.exit()
  }

  /** @internal */
  public purge() {
    while (this.stack.length) console.log(this.stack.shift());
  }

  /** @internal */
  public dispose(cb: () => void) {
    this.config.level = "silent"; // no more logs
    clearImmediate(this.postponed);
    this.postponed = setImmediate(() => {
      this.purge();
      cb();
    });
  }

  public debug(message: string, meta?: unknown) {
    this.print("debug", message, meta);
  }

  public info(message: string, meta?: unknown) {
    this.print("info", message, meta);
  }

  public warn(message: string, meta?: unknown) {
    this.print("warn", message, meta);
  }

  public error(message: string, meta?: unknown) {
    this.print("error", message, meta);
  }

  public child(ctx: Context) {
    return new BuiltinLogger({ ...this.config, ctx });
  }

  /**
   * @desc The argument used for instance created by .child() method
   * @see ChildLoggerProvider
   * */
  public get ctx() {
    return this.config.ctx;
  }

  /** @desc Measures the duration until you invoke the returned callback */
  public profile(message: string): () => void;
  public profile(options: ProfilerOptions): () => void;
  public profile(subject: string | ProfilerOptions) {
    const start = performance.now();
    return () => {
      const duration = performance.now() - start;
      const {
        message,
        severity = "debug",
        formatter = formatDuration,
      } = typeof subject === "object" ? subject : { message: subject };
      this.print(
        typeof severity === "function" ? severity(duration) : severity,
        message,
        formatter(duration),
      );
    };
  }
}
