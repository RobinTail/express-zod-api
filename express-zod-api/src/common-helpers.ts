import { Request } from "express";
import * as R from "ramda";
import { z } from "zod";
import { CommonConfig, InputSource, InputSources } from "./config-type";
import { contentTypes } from "./content-type";
import { OutputValidationError } from "./errors";
import { metaSymbol } from "./metadata";
import { AuxMethod, Method } from "./method";

/** @desc this type does not allow props assignment, but it works for reading them when merged with another interface */
export type EmptyObject = Record<string, never>;
export type EmptySchema = z.ZodObject<EmptyObject, "strip">;
export type FlatObject = Record<string, unknown>;

/** @link https://stackoverflow.com/a/65492934 */
type NoNever<T, F> = [T] extends [never] ? F : T;

/**
 * @desc Using module augmentation approach you can specify tags as the keys of this interface
 * @example declare module "express-zod-api" { interface TagOverrides { users: unknown } }
 * @link https://www.typescriptlang.org/docs/handbook/declaration-merging.html#module-augmentation
 * */
// eslint-disable-next-line @typescript-eslint/no-empty-object-type -- augmentation
export interface TagOverrides {}
export type Tag = NoNever<keyof TagOverrides, string>;

/** @see https://expressjs.com/en/guide/routing.html */
export const routePathParamsRegex = /:([A-Za-z0-9_]+)/g;
export const getRoutePathParams = (path: string): string[] =>
  path.match(routePathParamsRegex)?.map((param) => param.slice(1)) || [];

const areFilesAvailable = (request: Request): boolean => {
  const contentType = request.header("content-type") || "";
  const isUpload = contentType.toLowerCase().startsWith(contentTypes.upload);
  return "files" in request && isUpload;
};

export const defaultInputSources: InputSources = {
  get: ["query", "params"],
  post: ["body", "params", "files"],
  put: ["body", "params"],
  patch: ["body", "params"],
  delete: ["query", "params"],
};
const fallbackInputSource: InputSource[] = ["body", "query", "params"];

export const getActualMethod = (request: Request) =>
  request.method.toLowerCase() as Method | AuxMethod;

export const getInput = (
  req: Request,
  userDefined: CommonConfig["inputSources"] = {},
): FlatObject => {
  const method = getActualMethod(req);
  if (method === "options") return {};
  return (
    userDefined[method] ||
    defaultInputSources[method] ||
    fallbackInputSource
  )
    .filter((src) => (src === "files" ? areFilesAvailable(req) : true))
    .reduce<FlatObject>((agg, src) => Object.assign(agg, req[src]), {});
};

export const ensureError = (subject: unknown): Error =>
  subject instanceof Error ? subject : new Error(String(subject));

export const getMessageFromError = (error: Error): string => {
  if (error instanceof z.ZodError) {
    return error.issues
      .map(({ path, message }) =>
        (path.length ? [path.join("/")] : []).concat(message).join(": "),
      )
      .join("; ");
  }
  if (error instanceof OutputValidationError) {
    const hasFirstField = error.cause.issues[0]?.path.length > 0;
    return `output${hasFirstField ? "/" : ": "}${error.message}`;
  }
  return error.message;
};

/** Takes the original unvalidated examples from the properties of ZodObject schema shape */
export const pullExampleProps = <T extends z.SomeZodObject>(subject: T) =>
  Object.entries(subject.shape).reduce<Partial<z.input<T>>[]>(
    (acc, [key, schema]) => {
      const { _def } = schema as z.ZodType;
      return combinations(
        acc,
        (_def[metaSymbol]?.examples || []).map(R.objOf(key)),
        ([left, right]) => ({ ...left, ...right }),
      );
    },
    [],
  );

export const getExamples = <
  T extends z.ZodType,
  V extends "original" | "parsed" | undefined,
>({
  schema,
  variant = "original",
  validate = variant === "parsed",
  pullProps = false,
}: {
  schema: T;
  /**
   * @desc examples variant: original or parsed
   * @example "parsed" — for the case when possible schema transformations should be applied
   * @default "original"
   * @override validate: variant "parsed" activates validation as well
   * */
  variant?: V;
  /**
   * @desc filters out the examples that do not match the schema
   * @default variant === "parsed"
   * */
  validate?: boolean;
  /**
   * @desc should pull examples from properties — applicable to ZodObject only
   * @default false
   * */
  pullProps?: boolean;
}): ReadonlyArray<V extends "parsed" ? z.output<T> : z.input<T>> => {
  let examples = schema._def[metaSymbol]?.examples || [];
  if (!examples.length && pullProps && schema instanceof z.ZodObject)
    examples = pullExampleProps(schema);
  if (!validate && variant === "original") return examples;
  const result: Array<z.input<T> | z.output<T>> = [];
  for (const example of examples) {
    const parsedExample = schema.safeParse(example);
    if (parsedExample.success)
      result.push(variant === "parsed" ? parsedExample.data : example);
  }
  return result;
};

export const combinations = <T>(
  a: T[],
  b: T[],
  merge: (pair: [T, T]) => T,
): T[] => (a.length && b.length ? R.xprod(a, b).map(merge) : a.concat(b));

/**
 * @desc isNullable() and isOptional() validate the schema's input
 * @desc They always return true in case of coercion, which should be taken into account when depicting response
 */
export const hasCoercion = (schema: z.ZodTypeAny): boolean =>
  "coerce" in schema._def && typeof schema._def.coerce === "boolean"
    ? schema._def.coerce
    : false;

export const ucFirst = (subject: string) =>
  subject.charAt(0).toUpperCase() + subject.slice(1).toLowerCase();

export const makeCleanId = (...args: string[]) => {
  const byAlpha = R.chain((entry) => entry.split(/[^A-Z0-9]/gi), args);
  const byWord = R.chain(
    (entry) =>
      entry.replaceAll(/[A-Z]+/g, (beginning) => `/${beginning}`).split("/"),
    byAlpha,
  );
  return byWord.map(ucFirst).join("");
};

export const getTransformedType = R.tryCatch(
  <T>(schema: z.ZodEffects<z.ZodTypeAny, unknown, T>, sample: T) =>
    typeof schema.parse(sample),
  R.always(undefined),
);

/** @desc can still be an array, use Array.isArray() or rather R.type() to exclude that case */
export const isObject = (subject: unknown) =>
  typeof subject === "object" && subject !== null;

export const isProduction = R.memoizeWith(
  () => process.env.TSUP_STATIC as string, // eslint-disable-line no-restricted-syntax -- substituted by TSUP
  () => process.env.NODE_ENV === "production", // eslint-disable-line no-restricted-syntax -- memoized
);
