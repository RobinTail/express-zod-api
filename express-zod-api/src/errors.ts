import type { $ZodType, $ZodErrorTree } from "zod/v4/core";
import { z } from "zod/v4";
import { getMessageFromError } from "./common-helpers";
import { OpenAPIContext } from "./documentation-helpers";
import type { Method } from "./method";

/** @desc An error related to the wrong Routing declaration */
export class RoutingError extends Error {
  public override name = "RoutingError";
  public override readonly cause: { method: Method; path: string };

  constructor(message: string, method: Method, path: string) {
    super(message);
    this.cause = { method, path };
  }
}

/**
 * @desc An error related to the generating of the documentation
 * */
export class DocumentationError extends Error {
  public override name = "DocumentationError";
  public override readonly cause: string;

  constructor(
    message: string,
    {
      method,
      path,
      isResponse,
    }: Pick<OpenAPIContext, "path" | "method" | "isResponse">,
  ) {
    super(message);
    this.cause = `${
      isResponse ? "Response" : "Input"
    } schema of an Endpoint assigned to ${method.toUpperCase()} method of ${path} path.`;
  }
}

/** @desc An error related to the input and output schemas declaration */
export class IOSchemaError extends Error {
  public override name = "IOSchemaError";
}

export class DeepCheckError extends IOSchemaError {
  public override name = "DeepCheckError";

  constructor(public override readonly cause: $ZodType) {
    super("Found", { cause });
  }
}

/** @desc An error of validating the Endpoint handler's returns against the Endpoint output schema */
export class OutputValidationError extends IOSchemaError {
  public override name = "OutputValidationError";

  constructor(public override readonly cause: z.ZodError) {
    const prefixedPath = new z.ZodError(
      cause.issues.map(({ path, ...rest }) => ({
        ...rest,
        path: ["output", ...path],
      })),
    );
    super(getMessageFromError(prefixedPath), { cause });
  }
}

/** @desc An error of validating the input sources against the Middleware or Endpoint input schema */
export class InputValidationError extends IOSchemaError {
  public override name = "InputValidationError";
  public tree: $ZodErrorTree<unknown>;

  constructor(public override readonly cause: z.ZodError) {
    super(getMessageFromError(cause), { cause });
    this.tree = z.treeifyError(cause);
  }
}

/** @desc An error related to the execution or incorrect configuration of ResultHandler */
export class ResultHandlerError extends Error {
  public override name = "ResultHandlerError";

  constructor(
    /** @desc The error thrown from ResultHandler */
    public override readonly cause: Error,
    /** @desc The error being processed by ResultHandler when it failed */
    public readonly handled?: Error,
  ) {
    super(getMessageFromError(cause), { cause });
  }
}

export class MissingPeerError extends Error {
  public override name = "MissingPeerError";
  constructor(module: string) {
    super(
      `Missing peer dependency: ${module}. Please install it to use the feature.`,
    );
  }
}
