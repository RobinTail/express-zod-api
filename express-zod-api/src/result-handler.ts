import { Request, Response } from "express";
import { globalRegistry, z } from "zod/v4";
import {
  ApiResponse,
  defaultStatusCodes,
  NormalizedResponse,
} from "./api-response";
import { FlatObject, isObject } from "./common-helpers";
import { contentTypes } from "./content-type";
import { IOSchema } from "./io-schema";
import { ActualLogger } from "./logger-helpers";
import {
  DiscriminatedResult,
  ensureHttpError,
  getPublicErrorMessage,
  logServerError,
  normalize,
  ResultSchema,
} from "./result-helpers";

type Handler<RES = unknown> = (
  params: DiscriminatedResult & {
    /** null in case of failure to parse or to find the matching endpoint (error: not found) */
    input: FlatObject | null;
    /** can be empty: check presence of the required property using "in" operator */
    options: FlatObject;
    request: Request;
    response: Response<RES>;
    logger: ActualLogger;
  },
) => void | Promise<void>;

export type Result<S extends z.ZodType = z.ZodType> =
  | S // plain schema, default status codes applied
  | ApiResponse<S> // single response definition, status code(s) customizable
  | ApiResponse<S>[]; // Feature #1431: different responses for different status codes (non-empty, prog. check!)

export type LazyResult<R extends Result, A extends unknown[] = []> = (
  ...args: A
) => R;

export abstract class AbstractResultHandler {
  readonly #handler: Handler;
  /** @internal */
  public abstract getPositiveResponse(output: IOSchema): NormalizedResponse[];
  /** @internal */
  public abstract getNegativeResponse(): NormalizedResponse[];
  protected constructor(handler: Handler) {
    this.#handler = handler;
  }
  public execute(...params: Parameters<Handler>) {
    return this.#handler(...params);
  }
}

export class ResultHandler<
  POS extends Result,
  NEG extends Result,
> extends AbstractResultHandler {
  readonly #positive: POS | LazyResult<POS, [IOSchema]>;
  readonly #negative: NEG | LazyResult<NEG>;

  constructor(params: {
    /** @desc A description of the API response in case of success (schema, status code, MIME type) */
    positive: POS | LazyResult<POS, [IOSchema]>;
    /** @desc A description of the API response in case of error (schema, status code, MIME type) */
    negative: NEG | LazyResult<NEG>;
    /** @desc The actual implementation to transmit the response in any case */
    handler: Handler<z.output<ResultSchema<POS> | ResultSchema<NEG>>>;
  }) {
    super(params.handler);
    this.#positive = params.positive;
    this.#negative = params.negative;
  }

  /** @internal */
  public override getPositiveResponse(output: IOSchema) {
    return normalize(this.#positive, {
      variant: "positive",
      args: [output],
      statusCodes: [defaultStatusCodes.positive],
      mimeTypes: [contentTypes.json],
    });
  }

  /** @internal */
  public override getNegativeResponse() {
    return normalize(this.#negative, {
      variant: "negative",
      args: [],
      statusCodes: [defaultStatusCodes.negative],
      mimeTypes: [contentTypes.json],
    });
  }
}

export const defaultResultHandler = new ResultHandler({
  positive: (output) => {
    const responseSchema = z.object({
      status: z.literal("success"),
      data: output,
    });
    const { examples = [] } = globalRegistry.get(output) || {}; // pulling down:
    if (examples.length) {
      globalRegistry.add(responseSchema, {
        examples: examples.map((data) => ({
          status: "success" as const,
          data,
        })),
      });
    }
    return responseSchema;
  },
  negative: z
    .object({
      status: z.literal("error"),
      error: z.object({ message: z.string() }),
    })
    .example({
      status: "error",
      error: { message: "Sample error message" },
    }),
  handler: ({ error, input, output, request, response, logger }) => {
    if (error) {
      const httpError = ensureHttpError(error);
      logServerError(httpError, logger, request, input);
      return void response
        .status(httpError.statusCode)
        .set(httpError.headers)
        .json({
          status: "error",
          error: { message: getPublicErrorMessage(httpError) },
        });
    }
    response
      .status(defaultStatusCodes.positive)
      .json({ status: "success", data: output });
  },
});

/**
 * @deprecated Resist the urge of using it: this handler is designed only to simplify the migration of legacy APIs.
 * @desc Responding with array is a bad practice keeping your endpoints from evolving without breaking changes.
 * @desc This handler expects your endpoint to have the property 'items' in the output object schema
 * */
export const arrayResultHandler = new ResultHandler({
  positive: (output) => {
    const responseSchema =
      output instanceof z.ZodObject &&
      "items" in output.shape &&
      output.shape.items instanceof z.ZodArray
        ? output.shape.items
        : z.array(z.any());
    const meta = responseSchema.meta();
    if (meta?.examples?.length) return responseSchema; // has examples on the items, or pull down:
    const examples = (globalRegistry.get(output)?.examples || [])
      .filter(
        (example): example is { items: unknown[] } =>
          isObject(example) &&
          "items" in example &&
          Array.isArray(example.items),
      )
      .map((example) => example.items);
    if (examples.length) {
      globalRegistry
        .remove(responseSchema) // reassign to avoid cloning
        .add(responseSchema, { ...meta, examples });
    }
    return responseSchema;
  },
  negative: z.string().example("Sample error message"),
  handler: ({ response, output, error, logger, request, input }) => {
    if (error) {
      const httpError = ensureHttpError(error);
      logServerError(httpError, logger, request, input);
      return void response
        .status(httpError.statusCode)
        .type("text/plain")
        .send(getPublicErrorMessage(httpError));
    }
    if ("items" in output && Array.isArray(output.items)) {
      return void response
        .status(defaultStatusCodes.positive)
        .json(output.items);
    }
    throw new Error("Property 'items' is missing in the endpoint output");
  },
});
