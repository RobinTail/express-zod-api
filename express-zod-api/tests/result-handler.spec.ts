import { Response } from "express";
import createHttpError from "http-errors";
import { z } from "zod";
import {
  InputValidationError,
  arrayResultHandler,
  defaultResultHandler,
  ResultHandler,
} from "../src";
import { ResultHandlerError } from "../src/errors";
import { AbstractResultHandler, Result } from "../src/result-handler";
import {
  makeLoggerMock,
  makeRequestMock,
  makeResponseMock,
} from "../src/testing";

describe("ResultHandler", () => {
  describe("constructor()", () => {
    test("should support multiple response schemas depending on status codes", () => {
      const subject = new ResultHandler({
        positive: () => [
          { statusCode: 200, schema: z.literal("ok") },
          { statusCode: 201, schema: z.literal("kinda") },
        ],
        negative: [
          { statusCode: 400, schema: z.literal("error") },
          { statusCode: 500, schema: z.literal("failure") },
        ],
        handler: ({ response }) => {
          expectTypeOf(response).toEqualTypeOf<
            Response<"ok" | "kinda" | "error" | "failure">
          >();
          response.status(200).send("error");
        },
      });
      expect(subject).toBeInstanceOf(AbstractResultHandler);
    });
  });

  describe("getters", () => {
    test("should throw when result is defined as an empty array", () => {
      expect(() =>
        new ResultHandler({
          positive: () => [] as Result,
          negative: vi.fn(),
          handler: vi.fn(),
        }).getPositiveResponse(z.object({})),
      ).toThrow(
        new ResultHandlerError(
          new Error("At least one positive response schema required."),
        ),
      );
      expect(() =>
        new ResultHandler({
          positive: vi.fn(),
          negative: [] as Result,
          handler: vi.fn(),
        }).getNegativeResponse(),
      ).toThrow(
        new ResultHandlerError(
          new Error("At least one negative response schema required."),
        ),
      );
    });
  });

  const requestMock = makeRequestMock({
    method: "POST",
    url: "http://something/v1/anything",
  });

  describe.each([
    {
      subject: defaultResultHandler,
      name: "defaultResultHandler",
    },
    {
      subject: arrayResultHandler,
      name: "arrayResultHandler",
    },
  ])("$name", ({ subject }) => {
    test("Should handle generic error", () => {
      const responseMock = makeResponseMock();
      const loggerMock = makeLoggerMock();
      const error = new Error("Some error");
      subject.execute({
        error,
        input: { something: 453 },
        output: null,
        request: requestMock,
        response: responseMock,
        logger: loggerMock,
        options: {},
      });
      expect(loggerMock._getLogs().error).toMatchSnapshot();
      expect(responseMock._getStatusCode()).toBe(500);
      expect(responseMock._getHeaders()).toHaveProperty(
        "content-type",
        responseMock._isJSON() ? "application/json" : "text/plain",
      );
      expect(
        responseMock._isJSON()
          ? responseMock._getJSONData()
          : responseMock._getData(),
      ).toMatchSnapshot();
    });

    test("Should handle schema error", () => {
      const responseMock = makeResponseMock();
      const loggerMock = makeLoggerMock();
      subject.execute({
        error: new InputValidationError(
          new z.ZodError([
            {
              code: "invalid_type",
              message: "Expected string, got number",
              path: ["something"],
              expected: "string",
              input: 453,
            },
          ]),
        ),
        input: { something: 453 },
        output: null,
        options: {},
        request: requestMock,
        response: responseMock,
        logger: loggerMock,
      });
      expect(loggerMock._getLogs().error).toHaveLength(0);
      expect(responseMock._getStatusCode()).toBe(400);
      expect(responseMock._getHeaders()).toHaveProperty(
        "content-type",
        responseMock._isJSON() ? "application/json" : "text/plain",
      );
      expect(
        responseMock._isJSON()
          ? responseMock._getJSONData()
          : responseMock._getData(),
      ).toMatchSnapshot();
    });

    test("Should handle HTTP error", () => {
      const responseMock = makeResponseMock();
      const loggerMock = makeLoggerMock();
      subject.execute({
        error: createHttpError(404, "Something not found"),
        input: { something: 453 },
        output: null,
        options: {},
        request: requestMock,
        response: responseMock,
        logger: loggerMock,
      });
      expect(loggerMock._getLogs().error).toHaveLength(0);
      expect(responseMock._getStatusCode()).toBe(404);
      expect(responseMock._getHeaders()).toHaveProperty(
        "content-type",
        responseMock._isJSON() ? "application/json" : "text/plain",
      );
      expect(
        responseMock._isJSON()
          ? responseMock._getJSONData()
          : responseMock._getData(),
      ).toMatchSnapshot();
    });

    test("Should handle regular response", () => {
      const responseMock = makeResponseMock();
      const loggerMock = makeLoggerMock();
      subject.execute({
        error: null,
        input: { something: 453 },
        output: { anything: 118, items: ["One", "Two", "Three"] },
        options: {},
        request: requestMock,
        response: responseMock,
        logger: loggerMock,
      });
      expect(loggerMock._getLogs().error).toHaveLength(0);
      expect(responseMock._getStatusCode()).toBe(200);
      expect(responseMock._getJSONData()).toMatchSnapshot();
    });

    test("should forward output schema examples", () => {
      const apiResponse = subject.getPositiveResponse(
        z
          .object({
            str: z.string(),
            items: z.array(z.string()),
          })
          .meta({
            examples: [{ str: "test", items: ["One", "Two", "Three"] }],
          }),
      );
      expect(apiResponse).toHaveLength(1);
      expect(apiResponse[0].schema.meta()).toMatchSnapshot();
    });

    test("should generate negative response example", () => {
      const apiResponse = subject.getNegativeResponse();
      expect(apiResponse).toHaveLength(1);
      expect(apiResponse[0].schema.meta()).toMatchSnapshot();
    });
  });

  test("arrayResultHandler should fail when there is no items prop in the output", () => {
    const responseMock = makeResponseMock();
    const loggerMock = makeLoggerMock();
    const positiveSchema = arrayResultHandler
      .getPositiveResponse(
        z
          .object({ anything: z.number() })
          .meta({ examples: [{ anything: 118 }] }),
      )
      .pop()?.schema;
    expect(positiveSchema).toHaveProperty(["_zod", "def", "type"], "array");
    expect(positiveSchema).toHaveProperty(
      ["_zod", "def", "element", "_zod", "def", "type"],
      "any",
    );
    expect(() =>
      arrayResultHandler.execute({
        error: null,
        input: { something: 453 },
        output: { anything: 118 },
        options: {},
        request: requestMock,
        response: responseMock,
        logger: loggerMock,
      }),
    ).toThrowError(
      // delegated to LastResortHandler, having same format
      new Error("Property 'items' is missing in the endpoint output"),
    );
  });
});
