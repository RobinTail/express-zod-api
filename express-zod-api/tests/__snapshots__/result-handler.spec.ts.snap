// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ResultHandler > 'arrayResultHandler' > Should handle HTTP error 1`] = `"Something not found"`;

exports[`ResultHandler > 'arrayResultHandler' > Should handle generic error 1`] = `
[
  [
    "Server side error",
    {
      "error": InternalServerError({
        "cause": Error({
          "message": "Some error",
        }),
        "message": "Some error",
      }),
      "payload": {
        "something": 453,
      },
      "url": "http://something/v1/anything",
    },
  ],
]
`;

exports[`ResultHandler > 'arrayResultHandler' > Should handle generic error 2`] = `"Some error"`;

exports[`ResultHandler > 'arrayResultHandler' > Should handle regular response 1`] = `
[
  "One",
  "Two",
  "Three",
]
`;

exports[`ResultHandler > 'arrayResultHandler' > Should handle schema error 1`] = `"something: Expected string, got number"`;

exports[`ResultHandler > 'arrayResultHandler' > should forward output schema examples 1`] = `
{
  "examples": [
    [
      "One",
      "Two",
      "Three",
    ],
  ],
}
`;

exports[`ResultHandler > 'arrayResultHandler' > should generate negative response example 1`] = `
{
  "examples": [
    "Sample error message",
  ],
}
`;

exports[`ResultHandler > 'defaultResultHandler' > Should handle HTTP error 1`] = `
{
  "error": {
    "message": "Something not found",
  },
  "status": "error",
}
`;

exports[`ResultHandler > 'defaultResultHandler' > Should handle generic error 1`] = `
[
  [
    "Server side error",
    {
      "error": InternalServerError({
        "cause": Error({
          "message": "Some error",
        }),
        "message": "Some error",
      }),
      "payload": {
        "something": 453,
      },
      "url": "http://something/v1/anything",
    },
  ],
]
`;

exports[`ResultHandler > 'defaultResultHandler' > Should handle generic error 2`] = `
{
  "error": {
    "message": "Some error",
  },
  "status": "error",
}
`;

exports[`ResultHandler > 'defaultResultHandler' > Should handle regular response 1`] = `
{
  "data": {
    "anything": 118,
    "items": [
      "One",
      "Two",
      "Three",
    ],
  },
  "status": "success",
}
`;

exports[`ResultHandler > 'defaultResultHandler' > Should handle schema error 1`] = `
{
  "error": {
    "message": "something: Expected string, got number",
  },
  "status": "error",
  "tree": {
    "errors": [],
    "properties": {
      "something": {
        "errors": [
          "Expected string, got number",
        ],
      },
    },
  },
}
`;

exports[`ResultHandler > 'defaultResultHandler' > should forward output schema examples 1`] = `
{
  "examples": [
    {
      "data": {
        "items": [
          "One",
          "Two",
          "Three",
        ],
        "str": "test",
      },
      "status": "success",
    },
  ],
}
`;

exports[`ResultHandler > 'defaultResultHandler' > should generate negative response example 1`] = `
{
  "examples": [
    {
      "error": {
        "message": "Sample error message",
      },
      "status": "error",
    },
  ],
}
`;
